Templates-recent templates behavior - MAILPOET-5949

triple0t opened the pull request:
## Description

Continuation of https://github.com/mailpoet/mailpoet/pull/5975

In this PR, we update the recent template behavior by pulling emails sent from the new email editor and using its post content as an email preset (template) for new emails.

![Screenshot 2024-12-05 at 11 09 46 PM](https://github.com/user-attachments/assets/8aa7a572-1c61-49fd-b3b2-fb8c690e166e)
![Screenshot 2024-12-05 at 11 09 56 PM](https://github.com/user-attachments/assets/c7d6d66c-ef80-4dcd-b024-5f917418ac74)


## Code review notes

Please start review from https://github.com/mailpoet/mailpoet/pull/5980/commits/2bc05534c72a96ac6952d71bcd445eb9a4ea4e0b

## QA notes

For testing: You can add `draft` to this list https://github.com/mailpoet/mailpoet/blob/95a8b9bf0be8ea2bcabf5c810238609e1d765b6d/packages/js/email-editor/src/store/selectors.ts#L147
e.g.
```
status: 'publish,sent,draft',
```


## Linked PRs

Related to https://github.com/mailpoet/mailpoet/pull/5975

## Linked tickets

[MAILPOET-5949](https://mailpoet.atlassian.net/browse/MAILPOET-5949)

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:templates-recent-templates-behaviour)

_The latest successful build from `templates-recent-templates-behaviour` will be used. If none is available, the link won't work._

[MAILPOET-5949]: https://mailpoet.atlassian.net/browse/MAILPOET-5949?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

triple0t committed:
Add Selector to fetch mailpoet email editor post-types

MAILPOET-5949

triple0t committed:
Fetch emailPosts for use in template selection modal

MAILPOET-5949

triple0t committed:
Fix using sent email post content as new email editor page template

MAILPOET-5949

triple0t committed:
Fix an edge case where it might be possible for email content to have similar post-slug

MAILPOET-5949

triple0t committed:
Add support for email-general and email-simple templates

The previous implementation only checks for email-general templates

MAILPOET-5949

triple0t committed:
Add basic optimization for `usePreviewTemplates` to improve render performance

MAILPOET-5949

triple0t committed:
Refactor `SelectTemplateBody` and use memo and useEffect when required to improve render performance

MAILPOET-5949

triple0t committed:
Update template selector styles

MAILPOET-5949

triple0t committed:
Filter out empty email post-content

MAILPOET-5949

triple0t commented on packages/js/email-editor/src/components/template-select/select-modal.tsx:
```diff
@@ -1,34 +1,83 @@
 
 const BLANK_TEMPLATE = 'email-general';
 
+const TemplateCategories: Array< { name: TemplateCategory; label: string } > = [
```
We need to figure out what to do about this Template Category list. 
[WP_Block_Template](https://developer.wordpress.org/reference/classes/wp_block_template/) has an area property we can use to store the template category, but the field and value are not returned from the REST API. 

We would need to add another work-around [here](https://github.com/mailpoet/mailpoet/blob/b1cc53ec5bd0647e347a4ff281791b7e8148bad7/packages/php/email-editor/src/Engine/Templates/class-templates.php#L261-L287) by extending the `wp-template` endpoint. I'm not sure if there's value in doing that now or we should tackle it in this [ticket](https://mailpoet.atlassian.net/browse/MAILPOET-6356).

@costasovo, what do you think?

costasovo replied:
Thank you for pointing that out!

As for now we have only the two categories and we don't have problem figuring out what goes where so there is no rush.

I'm not sure the wp-template category and MailPoet selector categories are the same thing. In the selector user picks from template + initial content and I can imagine a template will be displayed in multiple categories because of a combination with a different pattern (or recent email content).
We will probably need some kind of registry and an extra endpoint for this in the future. I would wait with the implementation when we have a second category. The "recent emails" category is a specific automatic category.

triple0t replied:
Noted. Thank you for the feedback. 
We will leave it as it is then.

costasovo commented on packages/js/email-editor/src/store/selectors.ts:
```diff
@@ -140,6 +139,19 @@ export const getEditedEmailContent = createRegistrySelector(
+		const posts: EmailEditorPostType[] = select(
+			coreDataStore
+		).getEntityRecords( 'postType', 'mailpoet_email', {
+			per_page: 30, // show a maximum of 30 for now
```
Did you try with 30 posts? I am concerned that 30 previews in the modal will cause performance issues, as every preview is basically a small block canvas. But we can still go with 30 and adjust later...

triple0t replied:
Yeah, I did. It worked well.

costasovo commented on packages/js/email-editor/src/hooks/use-preview-templates.ts:
```diff
@@ -40,6 +40,53 @@ function setPostContentInnerBlocks(
+/**
+ * We are reusing the template CSS and mailpoet theme by fetching the template from
+ * the list of email editor available templates.
+ * Note: This function may need an update when https://mailpoet.atlassian.net/browse/MAILPOET-6335 is merged
```
Yeah. Great point! After MAILPOET-6335 there will be just global styles. Thank you for adding this workaround!

costasovo reviewed (changes requested):
Great job! It works nicely. 

When testing, I ran into an issue with the recent email preview.
I sent an email created with the Simple Light template and then in the recent emails category the preview rendered only the content part, but not the template.
See screenshot. We could also place the "Start from scratch" button fixed to the bottom right. I think it would look cleaner.

![Screenshot 2024-12-06 at 10 03 06](https://github.com/user-attachments/assets/0c360fd3-7af0-4bb2-b3d8-03b8a848d5f7)

triple0t committed:
Fix header and footer not showing for recent category email post templates

MAILPOET-5949

triple0t committed:
Ensure "Start from scratch" button sticks to the end of the modal

MAILPOET-5949

triple0t commented:
Thank you for the review, @costasovo 

> When testing, I ran into an issue with the recent email preview.
I sent an email created with the Simple Light template and then in the recent emails category the preview rendered only the content part, but not the template.

Fixed in https://github.com/mailpoet/mailpoet/pull/5980/commits/4588f92261101c3a89236eab8359478cb7f7cebf

> See screenshot. We could also place the "Start from scratch" button fixed to the bottom right. I think it would look cleaner.

Fixed in https://github.com/mailpoet/mailpoet/pull/5980/commits/ca4201f89b0e4f38248a41db386fdc6522f4ba8f

costasovo reviewed (commented):
@triple0t Thanks for the fix it works as expected!

I see there are conflicts with the trunk branch. It is most likely due to changes I made for email template swap functionality in https://github.com/mailpoet/mailpoet/commit/10b2c2b648af165898a01e8c550adea8347d9a49
This functionality ensures that when swapping a template all previews in the select modal use the content of the currently edited email so that user can see how the email will look like when they change the template.

I'm not sure if that behavior makes senses for recent emails category because it may end up showing just a bunch of same templates as many recent email will be based on the single template. I think we cold:
1) Disable recent emails category when swapping because all temlates should be listed anyway. We can detect it from previewContent being set 
2) Keep the recent email category with the currently edited email content. On select apply only template change
3) Keep the recent email category with the contents from sent emails. 
       3a) on change apply only template change
       3b) on change apply both template and content

I would personally go with the option 1) or 2). Any other ideas?

triple0t commented:
Oh, thanks for the ping, @costasovo 

Funny, I was focusing on fixing the merge conflict and wanted to message you about this upon completion. (You beat me to it)

> I'm not sure if that behavior makes senses for recent emails category because it may end up showing just a bunch of same templates as many recent email will be based on the single template. I think we cold:
> 
> 1. Disable recent emails category when swapping because all temlates should be listed anyway. We can detect it from previewContent being set
> 2. Keep the recent email category with the currently edited email content. On select apply only template change
> 3. Keep the recent email category with the contents from sent emails.
> 3a) on change apply only template change
> 3b) on change apply both template and content
> I would personally go with the option 1) or 2). Any other ideas?


Let's go with Option 2, as that is the current behavior.
If the need arises, we can change this later.

costasovo reviewed (dismissed):
@triple0t Thanks for fixing the conflict!

> Let's go with Option 2, as that is the current behavior.
If the need arises, we can change this later.

Makes sense! 👍

costasovo commented:
@triple0t I wanted to merge the PR, but I noticed a test started to fail after the latest change. Can you please take a look? I'll be able to look into it on Monday.

costasovo commented:
@triple0t Oh! And I realized I shouldn't merge it because it is on top of another PR. So, I'll leave merging to you.

triple0t commented:
Thank you, @costasovo 

Yeah, this PR is based on another.
I fixed the failing test in the parent PR and will rebase it after it is merged.

triple0t committed:
Fix acceptance test

MAILPOET-5949
